[BEAM-2865] Introduces FileIO.write() and uses it in AvroIO.

jkff opened the pull request:
https://issues.apache.org/jira/browse/BEAM-2865

http://s.apache.org/fileio-write

Depends on https://github.com/apache/beam/pull/3814.

R: @robertwb 
CC: @reuvenlax

jbonofre commented:
CC: @jbonofre

jkff commented:
OK, I think this is ready for review. I changed the API slightly and added a lot of documentation.

reuvenlax commented on sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileIO.java:
```diff
@@ -439,4 +652,553 @@ public void process(ProcessContext c) {
+     * window with the default trigger.
+     *
+     * @deprecated Avoid usage of this method: its effects are complex and it will be removed in
+     *     future versions of Beam. Right now it exists for compatibility with {@link WriteFiles}.
```
<!--new_thread; commit:28d55b3a9ab8a0cd84e7834b34c2c150edf871d7; resolved:0-->
There are specific semantic differences today (e.g. what is done with empty shards), so I'm not sure its use can be avoided. Maybe we'll be able to sort these semantic issues out in the future, but I don't think we are there yet.

reuvenlax commented on sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileIO.java:
```diff
@@ -21,38 +21,232 @@
+ *
+ * <ul>
+ *   <li><b>How many shards are generated per window:</b> This is controlled by <i>sharding</i>,
+ *       using {@link Write#withNumShards} or {@link Write#withSharding}. By default, every bundle
```
<!--new_thread; commit:28d55b3a9ab8a0cd84e7834b34c2c150edf871d7; resolved:0-->
Actually the default is runner specific. However if the runner does not do its own overriding, the default is one per bundle.

reuvenlax commented on sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileIO.java:
```diff
@@ -439,4 +652,553 @@ public void process(ProcessContext c) {
+    /**
+     * A policy for generating names for shard files using information in {@link FilenameContext}.
+     */
+    interface FilenamePolicy extends Serializable {
```
<!--new_thread; commit:28d55b3a9ab8a0cd84e7834b34c2c150edf871d7; resolved:0-->
Won't it get confusing having two FilenamePolicy interfaces?

reuvenlax commented on sdks/java/core/src/main/java/org/apache/beam/sdk/io/AvroIO.java:
```diff
@@ -1165,14 +1175,29 @@ public ResourceId apply(String input) {
+      FileIO.Write<DestinationT, UserT> write =
+          FileIO.<DestinationT, UserT>writeDynamic()
+              .by(new DestinationFnViaDynamicDestinations<>(dynamicDestinations))
+              .via(
```
<!--new_thread; commit:28d55b3a9ab8a0cd84e7834b34c2c150edf871d7; resolved:0-->
Will this cache sinks per destination, or constantly recreate sinks? If the latter., this might be a performance regression.

reuvenlax commented on sdks/java/core/src/main/java/org/apache/beam/sdk/io/AvroIO.java:
```diff
@@ -1165,14 +1175,29 @@ public ResourceId apply(String input) {
+                      input.getPipeline().getCoderRegistry()))
+              .withEmptyGlobalWindowDestination(dynamicDestinations.getDefaultDestination())
+              .withTempDirectory(tempDirectory)
+              .withSideInputs(dynamicDestinations.getSideInputs());
```
<!--new_thread; commit:28d55b3a9ab8a0cd84e7834b34c2c150edf871d7; resolved:0-->
Is this actually simpler? In particular things like getDestinationCoderWithDefault were handled automatically before, and now seems like users of FileIO need to know about this specifically.

reuvenlax commented on sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileIO.java:
```diff
@@ -439,4 +652,553 @@ public void process(ProcessContext c) {
+     * of {@link DefaultFilenamePolicy#DEFAULT_WINDOWED_SHARD_TEMPLATE}, which is {@code
+     * W-P-SSSSS-of-NNNNN}.
+     */
+    public static FilenamePolicy prefixAndWindowedShard(final String prefix) {
```
<!--new_thread; commit:28d55b3a9ab8a0cd84e7834b34c2c150edf871d7; resolved:0-->
the names of these methods are confusing - they don't tell me what they do.

reuvenlax commented on sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileIO.java:
```diff
@@ -21,38 +21,232 @@
+ * partition the elements into groups ("destinations"). Then elements will be grouped by
+ * destination, and {@link Write#to(SerializableFunction)} and the {@link
+ * Write#via(SerializableFunction)} will be applied separately within each group. Note that
+ * currently sharding can not be destination-dependent.
```
<!--new_thread; commit:28d55b3a9ab8a0cd84e7834b34c2c150edf871d7; resolved:0-->
Please explain this.

reuvenlax commented on sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileIO.java:
```diff
@@ -21,38 +21,232 @@
+ * Sink}, e.g. write different elements to Avro files in different directories with different
+ * schemas.
+ *
+ * <p>This feature is supported by {@link #writeDynamic}. Use {@link Write#by} to specify how to
```
<!--new_thread; commit:28d55b3a9ab8a0cd84e7834b34c2c150edf871d7; resolved:0-->
Are we giving up on anything more generic? BigQueryIO already has DynamicDestinations which we had planned on moving somewhere common (to use in other Sinks). However it now seems that we are moving away from DynamicDestinations, which means that every Sink will do its own thing for dynamic destinations (and BigQueryIO will remain different).

reuvenlax commented on sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileIO.java:
```diff
@@ -21,38 +21,232 @@
+ *       based on runner behavior, though at least 1 shard will always be produced for every
+ *       non-empty window. Note that setting a fixed number of shards can hurt performance: it adds
+ *       an additional {@link GroupByKey} to the pipeline.
+ *   <li><b>How the shards are named:</b> This is controlled by a {@link Write.FilenamePolicy}:
```
<!--new_thread; commit:28d55b3a9ab8a0cd84e7834b34c2c150edf871d7; resolved:0-->
More how the files are named, as a single shard might consist of amny files.

kennknowles commented:
Please be sure to include a change to `ParDoTranslation` to ensure this can be round-tripped through the portable model.

kennknowles commented:
(or whatever else is relevant)

jkff commented:
I went ahead and removed DynamicDestinations from Avro and it looked like a good idea to fold that into this PR because it almost makes the overall diff smaller.

reuvenlax commented on sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteBundlesToFiles.java:
```diff
@@ -222,7 +222,7 @@ public void processElement(ProcessContext c, BoundedWindow window) throws Except
       writer.write(c.element().getValue());
     } catch (Exception e) {
-      // Discard write result and close the write.
+      // Discard write result and close the writeToDynamic.
```
was this intended?

reuvenlax commented on sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIOTest.java:
```diff
@@ -647,7 +647,7 @@ public void testWritingFailsTableDoesNotExist() throws Exception {
                 KvCoder.of(ByteStringCoder.of(), IterableCoder.of(ProtoCoder.of(Mutation.class)))));
 
-    // Exception will be thrown by write.validate() when write is applied.
+    // Exception will be thrown by write.validate() when writeToDynamic is applied.
```
intended?

reuvenlax commented on sdks/java/io/hbase/src/test/java/org/apache/beam/sdk/io/hbase/HBaseIOTest.java:
```diff
@@ -355,7 +355,7 @@ public void testWriting() throws Exception {
     final String table = "TEST-TABLE-DOES-NOT-EXIST";
 
-    // Exception will be thrown by write.expand() when write is applied.
+    // Exception will be thrown by write.expand() when writeToDynamic is applied.
```
intended?

jkff commented:
Thanks, all comments addressed:

* Kept the default codepath of TextIO and AvroIO unchanged - i.e. still calling WriteFiles, for update compatibility. Instead, TextIO and AvroIO are simply offering sink() and it is tested in their unit tests, and I marked the dynamic-destinations methods of TextIO and AvroIO deprecated.
* Removed FilenameContext in favor of explicit arguments. Renamed FilenamePolicy to FileNaming. Revamped the way file naming works: see the relevant section in the javadoc - I believe this is the most convenient and fully general way: common cases are easy to do, custom naming is easy to do as well.
* Reverted irrelevant changes
* To your question about watermark - it's outside the scope of this PR, but the watermark will effectively follow the timestamp of the next file to be emitted, where the timestamp is the moment the file was first seen.

jkff commented:
Cham - could you take a look also? Reuven may or may not get a chance to before going on vacation but he said he'd be ok with just your LGTM at this stage.

chamikaramj commented:
Sure. Will take a look.

chamikaramj commented on sdks/java/core/src/test/java/org/apache/beam/sdk/io/AvroIOTest.java:
```diff
@@ -849,32 +899,105 @@ private void testDynamicDestinationsWithSharding(Sharding sharding) throws Excep
+            case RUNNER_DETERMINED:
+              break;
+            case WITHOUT_SHARDING:
+              write = write.withNumShards(1);
```
withoutSharding() ?

chamikaramj commented on sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java:
```diff
@@ -742,20 +742,21 @@ public void removeTemporaryFiles(Collection<ResourceId> filenames) throws IOExce
+        for (KV<FileResult<DestinationT>, ResourceId> entry : resultsToFinalFilenames) {
+          srcFiles.add(entry.getKey().getTempFilename());
+          dstFiles.add(entry.getValue());
+          LOG.info(
```
Do we want to log every file ?
